src/common/server.ts

```diff
@@ -203,3 +228,29 @@ export async function restartServer(
 
   return newLSClient;
 }
+
+async function getRuffVersion(settings: ISettings): Promise<[number, number, number]> {
+  let command;
+  if (settings.path[0]) {
+    command = `${settings.path[0]} ${RUFF_VERSION_ARGS.join(" ")}`;
+  } else {
+    command = `${
+      settings.interpreter[0]
+    } ${EXPERIMENTAL_SERVER_SCRIPT_PATH} ${RUFF_VERSION_ARGS.join(" ")}`;
+  }
+  traceInfo(`Checking Ruff's version with command: ${command}`);
+  const execCommand = promisify(exec);
+  const { stdout } = await execCommand(command, {
+    cwd: settings.cwd,
+  });
```

Comment:
What happens if executing the command fails? For example, because the ruff path is incorrect?

Reply:
If that happens, `stdout` will either output nothing or an invalid string, in which case the version will fail to parse and return as `[0, 0, 0]`

Reply:
(which means that we will never select the native server)

Reply:
What I understand what happens if e.g. the path to the ruff binary is invalid is that `execCommand` fails (the promise gets rejected). Is there any upstream code that catches the exception? Is defaulting to the `ruff-lsp` the right fallback in that case or should we instead show a message saying that the path is invalid?

Reply:
> Is there any upstream code that catches the exception?

Not that I'm aware of, but this was also the current behavior of the extension before this change - if the path is invalid, the server crashes. I'd definitely be in favor of handling this more gracefully, but I'd prefer doing that in a follow-up.

Reply:
I'm fine handling this in a follow up PR. But can we make sure that the output is about as good as for the current LSP?

```
2024-06-28 14:18:37.913 [info] [Error - 2:18:37 PM] Ruff client: couldn't create connection to server.
2024-06-28 14:18:37.913 [info] Launching server using command /home/micha/astral/ruff/target/debug/ruff2 failed. Error: spawn /home/micha/astral/ruff/target/debug/ruff2 ENOENT
2024-06-28 14:18:37.913 [error] Server: Start failed: Launching server using command /home/micha/astral/ruff/target/debug/ruff2 failed. Error: spawn /home/micha/astral/ruff/target/debug/ruff2 ENOENT
2024-06-28 14:18:37.913 [info]   22.461738414s  INFO ruff:main ruff_server::server::connection: Exit notification received. Server shutting down...
```

Reply:
Okay, I've tested this and it actually seems like we do just fall-back to `ruff-lsp`. I think that's fine for now, we can add the warning in a follow-up.